packages/protocol/contracts/L1/libs/LibProposing.sol

```diff
@@ -66,13 +71,32 @@ library LibProposing {
                 id: state.numBlocks,
                 timestamp: uint64(block.timestamp),
                 l1Height: uint64(block.number - 1),
-                gasLimit: input.gasLimit,
+                l2Basefee: 0, // will be set later
                 l1Hash: blockhash(block.number - 1),
                 mixHash: bytes32(block.prevrandao * state.numBlocks),
                 txListHash: input.txListHash,
                 txListByteStart: input.txListByteStart,
                 txListByteEnd: input.txListByteEnd,
-                beneficiary: input.beneficiary
+                gasLimit: input.gasLimit,
+                beneficiary: input.beneficiary,
+                treasure: resolver.resolve(config.chainId, "treasure", false)
+            });
+        }
+
+        // Calculate L2 EIP-1559 basefee per gas
+        //
+        // Note that we do not charge basefee * gaslimit on L1 as we do not
+        // know the actual gas used in the L2 block. If we charge the proposer
+        // here, the proposer may suffer a loss depends on how many enclosed
+        // transactions become invalid and are filtered out.
+        //
+        // On L2, EIP-1559's basefee will not be burned but send to a Taiko
+        // treasure address.
+        if (config.gasIssuedPerSecond != 0) {
+            (meta.l2Basefee, state.l2GasExcess) = LibL2Tokenomics.getL2Basefee({
+                state: state,
+                config: config,
+                gasLimit: input.gasLimit
```

Comment:
> So we still update the base fee based on the gasLimit instead of the gas_used, but only charge the proposer based on gas_used. Probably fine, but perhaps could lead to some abuse.

The basefee adjustment is based on the block's gas limit, not gasUsed as gasUsed is not known when blocks are proposed. But we don't charge proposer basefee* gasLimit, we charge the actual gasUsed * basefee on L2. On L1, no Ether will be collected from the proposer.